src/WebJobs.Script.WebHost/Security/KeyManagement/KubernetesSecretsRepository.cs

```diff
@@ -0,0 +1,216 @@
+                    }
+                }
+
+                if (hostSecret?.SystemKeys != null)
+                {
+                    foreach (var key in hostSecret.SystemKeys)
+                    {
+                        newKeys[$"{SystemKeyPrefix}{key.Name}"] = key.Value;
+                    }
+                }
+            }
+            else
+            {
+                // A function key is changing. Add all host and other functions keys
+                foreach (var key in currentKeys.Where(k => !k.Key.StartsWith($"{FunctionKeyPrefix}.{functionName}")))
+                {
+                    newKeys.Add(key.Key, key.Value);
+                }
+
+                var functionSecret = secrets as FunctionSecrets;
+                if (functionSecret?.Keys != null)
+                {
+                    foreach (var key in functionSecret.Keys)
+                    {
+                        newKeys[$"{FunctionKeyPrefix}{functionName}.{key.Name}"] = key.Value;
+                    }
+                }
+            }
+            return newKeys;
+        }
+
+        /// <summary>
+        /// no-op - allow stale secrets to remain
+        /// </summary>
+        public async Task PurgeOldSecretsAsync(IList<string> currentFunctions, ILogger logger)
+            => await Task.Yield();
+
+        /// <summary>
+        /// Runtime is not responsible for encryption so this code will never be executed.
+        /// </summary>
+        public Task WriteSnapshotAsync(ScriptSecretsType type, string functionName, ScriptSecrets secrets)
+            => throw new NotSupportedException();
+
+        /// <summary>
+        /// Runtime is not responsible for encryption so this code will never be executed.
+        /// </summary>
+        public Task<string[]> GetSecretSnapshots(ScriptSecretsType type, string functionName)
+            => throw new NotSupportedException();
+
+        private async Task<ScriptSecrets> ReadHostSecrets()
+        {
+            IDictionary<string, string> secrets = await _kubernetesClient.GetSecrets();
+
+            HostSecrets hostSecrets = new HostSecrets()
+            {
+                FunctionKeys = new List<Key>(),
+                SystemKeys = new List<Key>()
+            };
+
+            foreach (var pair in secrets)
+            {
+                if (pair.Key.StartsWith(MasterKey))
+                {
+                    hostSecrets.MasterKey = new Key("master", pair.Value);
+                }
+                else if (pair.Key.StartsWith(HostFunctionKeyPrefix))
+                {
+                    hostSecrets.FunctionKeys.Add(ParseKeyWithPrefix(HostFunctionKeyPrefix, pair.Key, pair.Value));
+                }
+                else if (pair.Key.StartsWith(SystemKeyPrefix))
+                {
+                    hostSecrets.SystemKeys.Add(ParseKeyWithPrefix(SystemKeyPrefix, pair.Key, pair.Value));
+                }
+            }
+
+            return hostSecrets?.MasterKey == null ? null : hostSecrets;
```

Comment:
I think there might be a bug in `KeyVaultSecretsRepository` [here ](https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs#L166) as I was looking at it as an example to make this one. 

SecretsManager assumes that if `ReadAsync` returned a non-null object for `HostSecrets`, then a `_master` key must exist. If this is the first time to use this repository on a brand new Secret (and I would assume KeyVault too, though I didn't test KeyVault) and you return a non-null `HostSecrets` without a `_master` key in it, all Admin operations will fail with a 500

/cc @alrod

Reply:
Yes, this does look like a bug.